refactor: split guest aborts / runtime errors

jakmeier opened the pull request:
This refactor makes a clear distinction between host and guest errors.

Guest errors are not actual errors from a runtime perspective. They
are completely normal outcomes of processing user defined code.

Host errors are usually due to bugs or a corrupted DB.

```rust
// new return type of VM::run
type VMResult = Result<VMOutcome, VMRunnerError>
```

`VMOutcome` contains a field `aborted` that contains the information
if the user code has finished with an error or not.

When something happens that prevents the VM runner from producing
a proper `VMOutcome`, the method will instead return `VMRunnerError`.

Resolves #7045.

jakmeier commented:
this PR is slightly better to review by commit by commit (but it touches many LOCs anyway...)

I think after this change, we could remove `AnyEq` because `AnyError` no lnoger needs to be comparable

matklad commented on runtime/near-vm-errors/src/lib.rs:
```diff
@@ -291,22 +292,21 @@ impl From<InconsistentStateError> for VMLogicError {
 
-impl From<VMLogicError> for VMError {
-    fn from(err: VMLogicError) -> Self {
+impl TryFrom<VMLogicError> for FunctionCallError {
```
I think this impl is load bearing, could you write a doc comment explaining the classification here (which error is ok, and which is err)

matklad commented on runtime/near-vm-errors/src/lib.rs:
```diff
@@ -7,18 +7,24 @@ use serde::{Deserialize, Serialize};
-    FunctionCallError(FunctionCallError),
-    /// Type erased error from `External` trait implementation.
-    ExternalError(AnyError),
+/// Fatal error occurred during VM operation.
```
`Fatal` is a bit unfortunate term here. It *feels* crisp enough to create a feeling of understanding in the reader, but actually explains nothing. 

Could you expand the definition a bit? That is, that this is error originating in runtime itself, rather than the contract, and that  the correct behavior is to either crash or create a challenge? 

The ultimate error handler here I think is 

https://github.com/near/nearcore/blob/0251fca578975beba51df7265af3b6cf7aabc0c6/nearcore/src/runtime/mod.rs#L515-L530

matklad commented on runtime/near-vm-runner/src/errors.rs:
```diff
@@ -1,7 +1,7 @@
 pub trait IntoVMError {
-    fn into_vm_error(self) -> VMError;
+    fn into_vm_error(self) -> Result<FunctionCallError, VMRunnerError>;
 }
```
In a follow-up, would be great to kill this trait as well, it makes the code less readable.

matklad commented on runtime/near-vm-runner/src/runner.rs:
```diff
@@ -1,16 +1,20 @@
-use crate::vm_kind::VMKind;
+/// Returned by VM::run method.
+///
+/// `VMOutcome` is a graceful completion of a VM execution, which may or
+/// may not have failed. `VMRunnerError` contains fatal errors that usually
+/// are panicked on by the caller.
```
This still sounds a bit confusing, is there perhaps a more useful wording? I think we want to underscore the differences between errors originating in the contract itself (contract divides by zero and dies) and in our runtime (runtime divides by zero and dies).

matklad commented on runtime/near-vm-runner/src/cache.rs:
```diff
@@ -73,11 +73,11 @@ fn cache_error(
 pub fn into_vm_result<T>(
     res: Result<Result<T, CompilationError>, CacheError>,
-) -> Result<T, VMError> {
+) -> Result<T, Result<FunctionCallError, VMRunnerError>> {
```
Hm, this feels like the nesting here is wrong. I'd expect it to read `Result<Result<T, FunctionCallError> , VMRunnerError>`, so that the call-site can ? the VM runner error?

jakmeier replied:
right, your nesting makes much more sense, changed it now

matklad reviewed (commented):
Yeah, this looks good on the first look through. I still need to do a line-by-line review here

jakmeier commented:
Tests are also updated now, CI will tell if I missed some code hidden behind feature flag combos... I was expecting this to be much worse tbh. Your refactoring @matklad of the tests is already paying dividends! It was quite quick to update >60 tests and check that the output is still equivalent.

I also changed the comments a bit, let me know if you think something is still unclear.

jakmeier commented:
not sure if I accidentally deleted some code here or if I encountered a bug in `UPDATE_EXPECT=1`

https://github.com/near/nearcore/pull/7815/commits/ebe7ee7399c4133c7a7ce698c1a560f3fa9a5f02

matklad commented on runtime/near-vm-runner/src/runner.rs:
```diff
@@ -105,7 +111,7 @@ pub trait VM {
         code_hash: &CryptoHash,
         cache: &dyn CompiledContractCache,
-    ) -> Option<VMError>;
+    ) -> Option<Result<FunctionCallError, VMRunnerError>>;
```
I think there's couple of problem here: 

* this should return CompilationError
* impls should use `?` to yeet VMRunnerError
* most critically, wasmtime should VMRunerError, not compilation error:

```diff
diff --git a/runtime/near-vm-errors/src/lib.rs b/runtime/near-vm-errors/src/lib.rs
index 95621e7c7..ab52b08c2 100644
--- a/runtime/near-vm-errors/src/lib.rs
+++ b/runtime/near-vm-errors/src/lib.rs
@@ -25,6 +25,9 @@ pub enum VMRunnerError {
     WasmUnknownError {
         debug_message: String,
     },
+    UnsupportedCompiler {
+        msg: String,
+    },
 }
 
 /// Permitted errors that cause a function call to fail gracefully.
@@ -151,7 +154,6 @@ pub enum CompilationError {
     CodeDoesNotExist { account_id: AccountId },
     PrepareError(PrepareError),
     WasmerCompileError { msg: String },
-    UnsupportedCompiler { msg: String },
 }
 
 #[derive(
@@ -394,9 +396,6 @@ impl fmt::Display for CompilationError {
             CompilationError::WasmerCompileError { msg } => {
                 write!(f, "Wasmer compilation error: {}", msg)
             }
-            CompilationError::UnsupportedCompiler { msg } => {
-                write!(f, "Unsupported compiler: {}", msg)
-            }
         }
     }
 }
@@ -419,6 +418,9 @@ impl fmt::Display for VMRunnerError {
             VMRunnerError::Nondeterministic(msg) => {
                 write!(f, "Nondeterministic error during contract execution: {}", msg)
             }
+            VMRunnerError::UnsupportedCompiler { msg } => {
+                write!(f, "Unsupported compiler {}", msg)
+            }
         }
     }
 }
diff --git a/runtime/near-vm-runner/src/runner.rs b/runtime/near-vm-runner/src/runner.rs
index 8776d95b7..c8858a4e5 100644
--- a/runtime/near-vm-runner/src/runner.rs
+++ b/runtime/near-vm-runner/src/runner.rs
@@ -5,7 +5,7 @@ use near_primitives::hash::CryptoHash;
 use near_primitives::runtime::fees::RuntimeFeesConfig;
 use near_primitives::types::CompiledContractCache;
 use near_primitives::version::ProtocolVersion;
-use near_vm_errors::{FunctionCallError, VMRunnerError};
+use near_vm_errors::{CompilationError, VMRunnerError};
 use near_vm_logic::types::PromiseResult;
 use near_vm_logic::{External, VMContext, VMOutcome};
 
@@ -117,7 +117,7 @@ pub trait VM {
         code: &[u8],
         code_hash: &CryptoHash,
         cache: &dyn CompiledContractCache,
-    ) -> Result<Option<FunctionCallError>, VMRunnerError>;
+    ) -> Result<Option<CompilationError>, VMRunnerError>;
 
     /// Verify the `code` contract can be compiled with this `VM`.
     ///
diff --git a/runtime/near-vm-runner/src/wasmer2_runner.rs b/runtime/near-vm-runner/src/wasmer2_runner.rs
index be0f5f731..8b7ceeba4 100644
--- a/runtime/near-vm-runner/src/wasmer2_runner.rs
+++ b/runtime/near-vm-runner/src/wasmer2_runner.rs
@@ -528,14 +528,14 @@ impl crate::runner::VM for Wasmer2VM {
         code: &[u8],
         code_hash: &near_primitives::hash::CryptoHash,
         cache: &dyn CompiledContractCache,
-    ) -> Result<Option<near_vm_errors::FunctionCallError>, VMRunnerError> {
-        let result = crate::cache::wasmer2_cache::compile_and_serialize_wasmer2(
+    ) -> Result<Option<near_vm_errors::CompilationError>, VMRunnerError> {
+        let outcome = crate::cache::wasmer2_cache::compile_and_serialize_wasmer2(
             code,
             code_hash,
             &self.config,
             cache,
-        );
-        let outcome = into_vm_result(result)?;
+        )
+        .map_err(VMRunnerError::CacheError)?;
         Ok(outcome.err())
     }
 
diff --git a/runtime/near-vm-runner/src/wasmer_runner.rs b/runtime/near-vm-runner/src/wasmer_runner.rs
index 84f40cf5a..d9ae80fee 100644
--- a/runtime/near-vm-runner/src/wasmer_runner.rs
+++ b/runtime/near-vm-runner/src/wasmer_runner.rs
@@ -327,14 +327,14 @@ impl crate::runner::VM for Wasmer0VM {
         code: &[u8],
         code_hash: &near_primitives::hash::CryptoHash,
         cache: &dyn CompiledContractCache,
-    ) -> Result<Option<FunctionCallError>, VMRunnerError> {
-        let result = crate::cache::wasmer0_cache::compile_and_serialize_wasmer(
+    ) -> Result<Option<CompilationError>, VMRunnerError> {
+        let outcome = crate::cache::wasmer0_cache::compile_and_serialize_wasmer(
             code,
             &self.config,
             code_hash,
             cache,
-        );
-        let outcome = into_vm_result(result)?;
+        )
+        .map_err(VMRunnerError::CacheError)?;
         Ok(outcome.err())
     }
 
diff --git a/runtime/near-vm-runner/src/wasmtime_runner.rs b/runtime/near-vm-runner/src/wasmtime_runner.rs
index 88f237ced..d52a45296 100644
--- a/runtime/near-vm-runner/src/wasmtime_runner.rs
+++ b/runtime/near-vm-runner/src/wasmtime_runner.rs
@@ -294,10 +294,10 @@ impl crate::runner::VM for WasmtimeVM {
         _code: &[u8],
         _code_hash: &CryptoHash,
         _cache: &dyn CompiledContractCache,
-    ) -> Result<Option<near_vm_errors::FunctionCallError>, VMRunnerError> {
-        Ok(Some(FunctionCallError::CompilationError(CompilationError::UnsupportedCompiler {
+    ) -> Result<Option<CompilationError>, VMRunnerError> {
+        Err(VMRunnerError::UnsupportedCompiler {
             msg: "Precompilation not supported in Wasmtime yet".to_string(),
-        })))
+        })
     }
 
     fn check_compile(&self, code: &[u8]) -> bool {
diff --git a/runtime/runtime/src/actions.rs b/runtime/runtime/src/actions.rs
index ed3629b1b..27c93c721 100644
--- a/runtime/runtime/src/actions.rs
+++ b/runtime/runtime/src/actions.rs
@@ -152,6 +152,9 @@ pub(crate) fn execute_function_call(
         VMRunnerError::WasmUnknownError { debug_message } => {
             panic!("Wasmer returned unknown message: {}", debug_message)
         }
+        VMRunnerError::UnsupportedCompiler { msg } => {
+            panic!("Wasmer returned unknown message: {}", msg)
+        }
     })
 }
 
```

jakmeier replied:
Agreed, this is how the errors should work here.

One problem, I noticed that `UnsupportedCompilerError` is part of the RPC schema:
https://github.com/near/nearcore/blob/26d1d7884a78f4e3759295559c0390f21f8aebac/chain/jsonrpc/res/rpc_errors_schema.json#L47-L56

Do we need to care about backward compatibility here? Or is this a dead error that never actually was stored anywhere?

matklad replied:
I *think* this is a dead error. I'd worry more about borsh than RPC schema, but I think we can be pretty sure that this is dead.

matklad replied:
I guess, we can leave `UnsupportedCompilerError` part of the change for a follow up as well?

matklad commented on runtime/near-vm-runner/src/wasmer2_runner.rs:
```diff
@@ -292,7 +293,7 @@ impl Wasmer2VM {
         mut import: Wasmer2Imports<'_, '_, '_>,
         method_name: &str,
-    ) -> Result<(), VMError> {
+    ) -> Result<(), Result<FunctionCallError, VMRunnerError>> {
```
Wrong nesting: 

```
Result<Result<(), FunctionCallError>, VMRunnerError>
```

Also, aling this with precompile's signature (in a follow up perhaps): both should either use Option or Result<(). My vote is for the latter.

jakmeier replied:
Conceptually, your nesting makes more sense. But in terms of yeeting `?`, it seems to be worse if I am not missing something....

The more common case is that we want to early return because of a `FunctionCallError`. This works with a `map_err(..)?` in the old (wrong) nesting. In the correct nesting, this is awkward because we want to early return on `Ok(Err(fn_call_error))` which is not something that `?` facilitates.

This translates the `?` I have now to something like this: 
```rust
// when foo() returns `Result<(), FunctionCallError>,`
let inner = match foo() {
    Ok(inner) => inner,
    Err(user_error => {
        return Ok(Err(user_error));
    }
};
```
Still trying to find a nicer solution here...

matklad replied:
Hm, yeah... I think that `Result<Result<(), FunctionCallError>, VMRunnerError>` is the conceptually right signature, and that we should use that troughout. Maybe even via `VMRunnerResult<T>` alias.

And yeah, this does require writing explicit matches for the inner result, but I think that's an Ok price to pay here (pun not intended). The inner thing is a value, it should be handled as value. Only the outer thing is semantically correct to through.

matklad reviewed (commented):
Run out of steam for today, but this is looking great!

matklad commented:
>not sure if I accidentally deleted some code here or if I encountered a bug in UPDATE_EXPECT=1



Bug in expect-test -- it doesn't correctly handlethe case when the same expect is re-used multiple times.

matklad commented on runtime/near-vm-runner/src/wasmer_runner.rs:
```diff
@@ -201,7 +195,11 @@ impl IntoVMError for wasmer_runtime::error::RuntimeError {
+    module: &Module,
+    import: &ImportObject,
+    method_name: &str,
+) -> Result<(), Result<FunctionCallError, VMRunnerError>> {
```
Flip Result order here

matklad commented on runtime/runtime/src/actions.rs:
```diff
@@ -115,7 +118,41 @@ pub(crate) fn execute_function_call(
     }
 
-    Ok(result)
+    // There are many specific errors but all of them are essentially unrecoverable.
+    // Some can be translated to the more general `RuntimeError`, which allows to pass
+    // the error up to the caller. For all other cases, panicking here is better
+    // than leaking the exact details further up.
+    // Note that this does not include errors caused by user code / input, those are
+    // stored in outcome.aborted.
```
Lets move from "recoverable" terminology in this comments as well

matklad commented on runtime/near-vm-errors/src/lib.rs:
```diff
@@ -25,6 +28,12 @@ pub enum VMRunnerError {
 
 /// Permitted errors that cause a function call to fail gracefully.
+///
+/// This error will be included in the merklize state on chain, it must be
+/// deterministically produces on all nodes.
```
Let's be a bit more precise here:

The _fact_ that this error occurred (one bit) will be included in the merklized state on chain. Specific details about error are not part of the chain, but they are stored in the database and returned on best-effort basis through RPC.

jakmeier replied:
corrected now, thanks for catching this!

matklad commented on runtime/near-vm-runner/src/runner.rs:
```diff
@@ -11,9 +11,15 @@ use near_vm_logic::{External, VMContext, VMOutcome};
+/// an error in the `aborted` field. But these are errors that may happen under
+/// normal circumstance. Such as when a smart contract code panics.
+/// Note that errors in `VMOutcome` are part of the state tracked on the
+/// blockchain. All validators must produce the same error deterministically.
```
>All validators must produce the same error deterministically.

I think this is not entirely correct: it's OK for validators to produce different errors, as long as they either both produce or not produce an error.

jakmeier replied:
also fixed

matklad commented:
Filed https://github.com/near/nearcore/issues/7818 for follow ups. 

I think I want only one change here before merge:

* In the comments, make it clear that details of errors are not important. Migh be a good idea to point to `PartialExecutionStatus`. 

I've collected the follow ups in https://github.com/near/nearcore/issues/7818.

jakmeier commented on runtime/near-vm-runner/src/tests/compile_errors.rs:
```diff
@@ -19,8 +19,7 @@ fn test_initializer_wrong_signature_contract() {
             expect![[r#"
-                VMOutcome: balance 4 storage_usage 12 return data None burnt gas 0 used gas 0
-                Err: PrepareError: Error happened while deserializing the module.
+                VMOutcome: balance 4 storage_usage 12 return data None burnt gas 0 used gas 0 failed with PrepareError: Error happened while deserializing the module.
```
Is multi line formatting in a `Debug` impl acceptable? Or would you suggest to no longer use that at all?

matklad replied:
I think multiline debug is not great, but also, our current Debug for VMOutcome is already not great, so no harm that way. 

But yeah, I think I was wrong to rely on Debug here at all, tests should just hand-craft the nice formatting.

jakmeier replied:
ok, updated
Not only is it nicely on two lines again, it is now 100% back to the same input / output as before and hence the commit doesn't touch the actual tests at all! :D
